ivy/functional/backends/tensorflow/experimental/statistical.py

```diff
@@ -160,6 +161,125 @@ def corrcoef(
     return cor
 
 
+def set_value(ret):
+    if isinstance(ret, list):
+        return set_value(ret[0])
+    elif isinstance(ret, np.ndarray):
+        return set_value(ret[0])
+    else:
+        return ret
+
+
+def _nanmedian_helper(input, axis=None, keepdims=False):
+    dtype = input.dtype
+    temp = tf.cast(input, tf.float64)
+    num_dim = tf.rank(temp)
+    keepdim_shape = tf.shape(temp)
+    q = 50.0
+
+    if axis is None:
+        if tf.reduce_all(tf.math.is_nan(temp)).numpy():
+            temp = tf.reshape(temp, shape=(1, -1))
+        else:
+            temp = temp[~tf.math.is_nan(temp)]
+
+        ret = tfp.stats.percentile(
+            temp,
+            q,
+            axis=axis,
+            interpolation="midpoint",
+            keepdims=keepdims,
+        )
+        ret = tf.cast(ret, dtype=dtype)
+        return ret
+
+    if isinstance(axis, tuple):
+        if len(axis) == 1:
+            axis = axis[0]
+
+    if isinstance(axis, int):
+        axis = [axis]
+    if isinstance(axis, tuple):
+        axis = list(axis)
+
+    for i in axis:
+        keepdim_shape = tf.tensor_scatter_nd_update(keepdim_shape, [[i]], [1])
+    axis = [num_dim + x if x < 0 else x for x in axis]
+    axis.sort()
+    dimension = tf.size(temp.shape)
+    while tf.size(axis) > 0:
+        axis1 = axis[0]
+        for axis2 in range(axis1 + 1, dimension):
+            temp = tf.transpose(
+                temp,
+                perm=tf.tensor_scatter_nd_update(
+                    tf.range(tf.rank(temp)), [[axis1], [axis2]], [axis2, axis1]
+                ),
+            )
+            axis1 = axis2
+        axis = [x - 1 for x in axis]
+        axis.pop(0)
+        dimension = dimension - 1
+    temp = tf.reshape(
+        temp, shape=tf.concat([tf.shape(temp)[: (dimension - len(axis))], [-1]], axis=0)
+    )
+    temp = tf.cast(temp, dtype)
+
+    tensor = tf.reshape(temp, shape=(1, -1))
+    shape = temp.shape
+    dim = temp.ndim
+    slice_size = shape[len(shape) - 1]
+    num_jumps = 1
+    result = []
+
+    if slice_size == 1:
+        if dim == 2 and input.shape[0] == 1:
+            return tensor
+        if dim > 2 and input.shape[0] == 1:
+            return tf.reshape(tensor, shape=input.shape)
+
+        tensor = tf.reshape(tensor, shape=shape[:-1])
+        return tensor
+
+    i = dim
+    while i > 1:
+        num_jumps *= shape[len(shape) - i]
+        i -= 1
+
+    for i in range(num_jumps):
+        start = i * slice_size
+        end = (i + 1) * slice_size
+        arr = tensor[:, start:end]
+        if tf.reduce_all(tf.math.is_nan(arr)).numpy():
+            arr = tf.reshape(arr, shape=(1, -1))
+        else:
+            arr = arr[~tf.math.is_nan(arr)]
+
+        ret = tfp.stats.percentile(
+            arr, q, axis=None, interpolation="midpoint", keepdims=keepdims
+        ).numpy()
+        if keepdims:
+            ret = set_value(ret)
```

Comment:
you are right about the row-wise median calculation 🙂, but for high dimensional inputs like 4D, 3D. or others, it returns in the form of an array like: [12.00] or [[12.00]], so here we only need the value of `ret` which is 12.00 because if we have this `ret` in form of an array, the line :` result = tf.reshape(result, shape=shape[:-1]).numpy()` will raise an error of something can't reshape eager Tensor like. so for the safe side, this helper is needed 😅. I hope it makes it clear 🙂

Reply:
Just a small correction: if you are specifying `keepdims=True` then this case happens which is about nested list generation from the percentile, by the definition of the keepdims means if it is true then it keeps the `last dimension` so whenever you set the `keepdims=Ture`, this case happens.
Here is an example: 
x = `[[np.nan, 12.0, np.nan, np.nan],[2., 3., 0.4, np.nan]]`
axis=`None`
keepdims=`True`

now for every iteration you print `ret`, you'll get:


arr =  tf.Tensor([2.], shape=(1,), dtype=float32)
ret =  [2.]
arr =  tf.Tensor([12.  3.], shape=(2,), dtype=float32)
ret =  [7.5]
arr =  tf.Tensor([0.4], shape=(1,), dtype=float32)
ret =  [0.4]
arr =  tf.Tensor([[nan nan]], shape=(1, 2), dtype=float32)
ret =  [[nan]

now we can't directly use `ret`, so we check if keepdims as `True` so it may be nested because of its nature then we remove its nesting. 

Now the question arises why the arr is in nested form, the answer is if you check this line: `temp = tf.reshape(temp, shape=tf.concat([tf.shape(temp)[: (dimension - len(axis))], [-1]], axis=0))`
the line is used for making the input according to our axis because we can only use axis=`None` in the `percentile` parameter and the reason is that the axis=0/1 in percentile works somehow differently than our desired functionality (which is calculating the median), we need the functionality of calculating the `median` with some fixed parameters of `percentile` so we need to make it (axis=None) here and for making it axis=None then we need to make it compatible for percentile with respect to calculating the `median`. 
if you debug the code you'll find temp's shape is changed according to the reason I explained above and that reshaped temp will be `flattened` which makes it like this: `[[ nan  2.  12.   3.   nan  0.4  nan  nan]]` with `shape=(1,8)` and then it moves further for calculation.

I hope it makes sense 🙂.

Reply:
> x = [[np.nan, 12.0, np.nan, np.nan],[2., 3., 0.4, np.nan]]
axis=None
keepdims=True

In the case of `axis=None`, I think `nanmedian` is calculated over all the elements of the input tensor. Hence, the median would be a scalar regardless of whether `keepdims` is `True` or `False`. So the question of nested lists is irrelevant here if my understanding is correct. 
For the case when `axis` is *not* `None`, you are transposing and then reshaping the input in a form that allows you to calculate the median row-wise. Hence, regardless of the dimension of the input tensor (`4D`,`5D` etc..), you'll alqays end up with a 2D tensor. Consequently, calculating the median row-wise for a 2D tensor *should* result in a 1D output. 
And so I don see the point of having the recursive `set_value()` helper here. Hope that makes sense 🙂

Reply:
when we have axis = `None` from the user side, we will not touch the working of `set_value` and `transposing`. the purpose of using `set_value` is for axis _not__ `None`. 
about not using recursive `set_value()` it makes sense now that it is good to not use it because we have always had 1D or 2D in `ret` and `ret` only contains `2D` when we have `nans`  else it yet returns `1D`.  we need to have multiple checks in set_value function for this.

Reply:
> we need to have multiple checks in set_value function for this. like checking if the ret is 2D or 1D and whether it is a list or an np.ndarray I guess,

I think we can remove this function altogether and simply have `ret=ret[0]` if that makes sense 🙂

Reply:
it only works for 1D but for 2D it returns 1D if we do so.

Reply:
yeah but when why we have `[[nan]]` as an output 🤔 ? Can you give an example

Reply:
there is a specific check: `if tf.reduce_all(tf.math.is_nan(arr)).numpy()` inside the jump loop wherein you are checking if all the values are nans, if they are nans then we don't need to use this line: `arr = arr[~tf.math.is_nan(arr)]` because it will remove all the `nans` but we need one nan for working on it if it make sense 🙂.

like about the above example: 
arr = tf.Tensor([2.], shape=(1,), dtype=float32)
ret = [2.]
arr = tf.Tensor([12. 3.], shape=(2,), dtype=float32)
ret = [7.5]
arr = tf.Tensor([0.4], shape=(1,), dtype=float32)
ret = [0.4]
_On this line it will not use the `tf.math.is_nan`, it just flatten it and need one `nan` because of the nature of nanmedian functionality_
arr = tf.Tensor([[nan nan]], shape=(1, 2), dtype=float32)
ret = [[nan] (this was a typo sorry for that 😅) it should be ret = [[nan]].

Reply:
if you run this, input = `[[2.0, ivy.nan, ivy.nan, 5.6, 7.8],[ivy.nan, ivy.nan, ivy.nan,ivy.nan,ivy.nan]]`
this will produce 5.6 and [[nan]], but here we need output like: [[5.6, nan]]. this is just one case and there are multiple cases with multiple values where we need to attach them and make it specific shape output as numpy's output shape, it wouldn't be that easy to generate a same shape output when you have some values are in shape=() or in shape=(1,1) I guess so.

Reply:
oh yeah makes sense. In that case, cant we simply do the following: 
```python
if keepdims:
   ret = tf.squeeze(ret) 
```
this will achieve the same purpose and is more intuitive I think 🙂

Reply:
yup, good idea!
I updated the code, if anything seems incorrect let me know 🙂